lib/ecto/exceptions.ex

```diff
@@ -81,22 +81,45 @@ defmodule Ecto.InvalidChangesetError do
   defexception [:action, :changeset]
 
   def message(%{action: action, changeset: changeset}) do
+    changes = extract_changes(changeset)
+    errors = Ecto.Changeset.traverse_errors(changeset, & &1)
+
     """
     could not perform #{action} because changeset is invalid.
 
-    * Changeset changes
+    Changes
+
+    #{pretty changes}
 
-    #{inspect changeset.changes}
+    Params
 
-    * Changeset params
+    #{pretty changeset.params}
 
-    #{inspect changeset.params}
+    Errors
 
-    * Changeset errors
+    #{pretty errors}
 
-    #{inspect changeset.errors}
+    Changeset
+
+    #{pretty changeset}
     """
   end
+
+  defp pretty(term) do
+    inspect(term, pretty: true)
+    |> String.split("\n")
+    |> Enum.map_join("\n", &"    " <> &1)
+  end
+
+  defp extract_changes(%Ecto.Changeset{changes: changes}) do
```

Comment:
If the changeset is marked as `action: :delete`, should we maybe remove it from showing?